src/libs/globbing/globbing.c

```diff
@@ -0,0 +1,212 @@
+/**
+ * @file
+ *
+ * @brief Library for performing globbing on keynames.
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <kdb.h>
+#include <kdbglobbing.h>
+#include <kdbhelper.h>
+
+#include <ctype.h>
+#include <fnmatch.h>
+#include <stdlib.h>
+#include <string.h>
+
+static size_t strcnt (const char * s, char c)
+{
+	size_t count = 0;
+	while ((s = strchr (s + 1, c)) != NULL)
+	{
+		count++;
+	}
+	return count;
+}
+
+static char * elektraToFnmatchGlob (char * pattern)
+{
+	char * ptr = pattern;
+	while ((ptr = strchr (ptr, '/')) != NULL)
+	{
+		++ptr;
+		if ((*(ptr + 1) == '/' || *(ptr + 1) == '\0') && (*ptr == '#' || *ptr == '_'))
+		{
+			*ptr = '*'; // replace /#/ and /_/ with /*/
+		}
+	}
+	return pattern;
+}
+
+/**
+ * @brief checks whether the given string is a valid elektra array item name
+ *
+ * Valid array items consist of a '#' followed by <code>n</code> underscores ('_'),
+ * followed by <code>n+1</code> digits ('0'-'9'). Additionally the digits describe a
+ * valid 32-bit Integer (i.e. <code>0 <= x < 2^32</code>).
+ *
+ * @param name the string to check
+ * @retval true if @p name is a valid array item
+ * @retval false otherwise
+ */
+bool isArrayName (const char * name) // TODO: move? definitely useful elsewhere, e.g. reference plugin, libease, etc.
+{
+	if (name == NULL || *name != '#')
+	{
+		return false;
+	}
+
+	++name;
+
+	int underscores = 0;
+	while (*name == '_')
+	{
+		name++;
+		underscores++;
+	}
+
+	int digits = 0;
+	while (isdigit ((unsigned char) *name))
+	{
+		name++;
+		digits++;
+	}
+
+	return digits + underscores + 2 <= ELEKTRA_MAX_ARRAY_SIZE && digits == underscores + 1;
+}
+
+/**
+ * @brief checks whether a given Key matches a given globbing pattern
+ *
+ * The globbing patterns for this function are a superset of those from fnmatch(3)
+ * used with the FNM_PATHNAME flag:
+ * <ul>
+ * 	<li> '*' matches any series of characters other than '/'</li>
+ * 	<li> '?' matches any single character except '/' </li>
+ * 	<li> '#', when used as "/#/" (or "/#" at the end of @p pattern), matches a valid array item </li>
+ * 	<li> '_', when used as "/_/"(or "/_" at the end of @p pattern), matches a key part that is <b>not</b> a valid array item </li>
+ * 	<li> if the pattern ends with "&#47;**", matching key names may contain arbitrary suffixes </li>
```

Comment:
Thank you, very nice feature! I was just thinking this would be great. Does this also makes sense within the key name?

And what about using `__` so that it is clear that it is an Elektra extension?

About how it is written down here: Why using '&#47;' instead of /?

Reply:
Or using _* would be another option. I am also not completely against **. It is, however, already used in zsh. So if we use **, it should have the same meaning.

Reply:
Yes you are right we should use `__` or possibly just `...`, because our implementation does not support using this syntax anywhere but at the end of the pattern. To support this behaviour inside the pattern, we would need a non-standard implementation of `fnmatch(3)`. (It might be possible with the GNU extension `FNM_EXTMATCH`)

> Why using '/' instead of /?
If you wanted to know why I used `&#47;`, that is because my IDE (or more specifically clang-tidy) complains about the use of `/*` inside of block comments.